### PR TITLE
Add support for per-job target manager timeouts

### PR DESCRIPTION
**Pull request by fabriziocucci**
This introuces per-job acquire and release timeouts.

### CODE REVIEW

**Comment by fabriziocucci** [comment]
While this seems to be the most straightforward solution, there is currently a circular dependency to solve:
```
$ ./run_lint.sh
+ export GO111MODULE=on
+ GO111MODULE=on
+ curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh
++ go env GOPATH
+ sh -s -- -b /Users/fabriziocucci/go/bin latest
golangci/golangci-lint info checking GitHub for tag 'latest'
golangci/golangci-lint info found version: 1.36.0 for v1.36.0/darwin/amd64
golangci/golangci-lint info installed /Users/fabriziocucci/go/bin/golangci-lint
+ go mod download
+ golangci-lint run --disable typecheck --enable deadcode --enable varcheck --enable staticcheck
ERRO Running error: context loading failed: failed to load packages: failed to load with go/packages: internal error: go list gives conflicting information for package github.com/facebookincubator/contest/pkg/job [github.com/facebookincubator/contest/pkg/job.test] 

$ go list github.com/facebookincubator/contest/pkg/job
package github.com/facebookincubator/contest/pkg/job
        imports github.com/facebookincubator/contest/pkg/event/testevent
        imports github.com/facebookincubator/contest/pkg/target
        imports github.com/facebookincubator/contest/pkg/job: import cycle not allowed
```

IMHO it's more a design issue that has been there all long and cleverly hidden by the package `pkg/types/types.go`, i.e. if `type JobID uint64` was part of `github.com/facebookincubator/contest/pkg/job` this would have surfaced earlier on.

I'll discuss this with @tfg13 and see what we can come up with.

**Comment by fabriziocucci** [comment]
Implemented workaround to keep hiding the problem! 😅

File: plugins/targetlocker/dblocker/dblocker.go [hunk 1]
```diff
@@ -276,14 +273,14 @@ func (d *DBLocker) Lock(jobID types.JobID, targets []*target.Target) error {
 	if len(targets) == 0 {
 		return nil
 	}
-
-	_, err := d.handleLock(int64(jobID), targetIDList(targets), uint(len(targets)), d.lockTimeout, false)
+	lockTimeout := jobTargetManagerAcquireTimeout + config.LockRefreshTimeout
```
Comment (tfg13) [inline plugins/targetlocker/dblocker/dblocker.go:276]: Aaah you are adding the refresh timeout here. I am fine with this, can you put a note this is added internally into the API doc (in locker.go) so people know

File: plugins/targetlocker/dblocker/dblocker.go [hunk 2]
```diff
@@ -294,8 +291,8 @@ func (d *DBLocker) TryLock(jobID types.JobID, targets []*target.Target, limit ui
 	if len(targets) == 0 || limit == 0 {
 		return nil, nil
 	}
-
-	return d.handleLock(int64(jobID), targetIDList(targets), limit, d.lockTimeout, true)
+	lockTimeout := jobTargetManagerAcquireTimeout + config.LockRefreshTimeout
```
Comment (tfg13) [inline plugins/targetlocker/dblocker/dblocker.go:294]: Same

File: plugins/targetmanagers/targetlist/targetlist.go [hunk 3]
```diff
@@ -84,13 +86,14 @@ func (t TargetList) ValidateReleaseParameters(params []byte) (interface{}, error
...
 // Acquire implements contest.TargetManager.Acquire
-func (t *TargetList) Acquire(ctx context.Context, jobID types.JobID, parameters interface{}, tl target.Locker) ([]*target.Target, error) {
+func (t *TargetList) Acquire(ctx context.Context, jobID types.JobID, jobTargetManagerAcquireTimeout time.Duration, parameters interface{}, tl target.Locker) ([]*target.Target, error) {
 	acquireParameters, ok := parameters.(AcquireParameters)
 	if !ok {
 		return nil, fmt.Errorf("Acquire expects %T object, got %T", acquireParameters, parameters)
 	}
 
-	if err := tl.Lock(jobID, acquireParameters.Targets); err != nil {
+	lockTimeout := jobTargetManagerAcquireTimeout + config.LockRefreshTimeout
```
Comment (tfg13) [inline plugins/targetmanagers/targetlist/targetlist.go:95]: You moved this inside the locker, right? (Adding the refresh timeout)

File: tests/plugins/targetlocker/dblocker/dblocker_test.go [hunk 4]
```diff
@@ -23,6 +23,7 @@ import (
 
 var (
 	jobID = types.JobID(123)
+	jobRargetManagerAcquireTimeout = 5 * time.Minute
```
Comment (tfg13) [inline tests/plugins/targetlocker/dblocker/dblocker_test.go:26]: Typo (`Rarget`) (and then copy-paste in the rest of the file :P)

Comment (tfg13) [inline tests/plugins/targetlocker/dblocker/dblocker_test.go:26]: The timeout for all these tests should be 2 seconds for initial + 2 seconds for refresh, see line 42.
Does this compile?

**Review by tfg13 — CHANGES_REQUESTED** [review]
LGTM overall, requesting changes because I think the dblocker unit test won't compile.

File: plugins/targetlocker/dblocker/dblocker.go (hunk 1 above)
Comment (fabriziocucci) [reply plugins/targetlocker/dblocker/dblocker.go:276]: Reverted this, it was probably an experiment leftover.

File: plugins/targetlocker/dblocker/dblocker.go (hunk 2 above)
Comment (fabriziocucci) [reply plugins/targetlocker/dblocker/dblocker.go:294]: Reverted this, it was probably an experiment leftover.

File: tests/plugins/targetlocker/dblocker/dblocker_test.go (hunk 4 above)
Comment (fabriziocucci) [reply tests/plugins/targetlocker/dblocker/dblocker_test.go:26]: Yep, I realised this now, changing it.

File: plugins/targetmanagers/targetlist/targetlist.go (hunk 3 above)
Comment (fabriziocucci) [reply plugins/targetmanagers/targetlist/targetlist.go:95]: Reverted this, it was probably an experiment leftover.

File: tests/plugins/targetlocker/dblocker/dblocker_test.go (hunk 4 above)
Comment (fabriziocucci) [reply tests/plugins/targetlocker/dblocker/dblocker_test.go]: I've fixed the lock timeout to be 2 seconds but, apparently, it doesn't really matter if the refresh timeout is left unchanged (e.g. 5 minutes from the default config).

File: pkg/target/locker.go [hunk 5]
```diff
@@ -33,15 +33,15 @@ type Locker interface {
 	// leaves the existing locks untouched in case of conflicts.
 	// Locks are reentrant, locking existing locks (with the same owner)
 	// extends the deadline.
-	Lock(types.JobID, []*Target) error
+	Lock(jobID types.JobID, jobTargetManagerAcquireTimeout time.Duration, targets []*Target) error
```
Comment (tfg13) [inline pkg/target/locker.go:36]: Can you rename this here, the Lock API shouldn't know/care if this is used by a target manager or something else. Just name it livetime or duration or something like that.

Also please fix the sentence 3 lines above that says the timeout is set at construction time :)

(same below for `TryLock`)

File: pkg/runner/job_runner.go [hunk 6]
```diff
@@ -104,7 +104,7 @@ func (jr *JobRunner) Run(j *job.Job) ([][]*job.Report, []*job.Report, error) {
 				// the Acquire semantic is synchronous, so that the implementation
 				// is simpler on the user's side. We run it in a goroutine in
 				// order to use a timeout for target acquisition.
-				targets, err := bundle.TargetManager.Acquire(j.StateCtx, j.ID, bundle.AcquireParameters, tl)
+				targets, err := bundle.TargetManager.Acquire(j.StateCtx, j.ID, j.TargetManagerAcquireTimeout, bundle.AcquireParameters, tl)
```
Comment (tfg13) [inline pkg/runner/job_runner.go:107]: I think this should be `j.TargetManagerAcquireTimeout + config. LockRefreshTimeout`, right? Just `j.TargetManagerAcquireTimeout` will race with the `Lock` below, so we need a bit more

**Review by tfg13 — COMMENTED** [review]
2 minor things to fix, then gtg

File: pkg/target/locker.go (hunk 5 above)
Comment (fabriziocucci) [reply pkg/target/locker.go:36]: Aliegned comments.

File: pkg/runner/job_runner.go (hunk 6 above)
Comment (fabriziocucci) [reply pkg/runner/job_runner.go:107]: Sounds fair.

**Review by tfg13 — APPROVED** [review]

**Pull request merged**